clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -134,6 +135,7 @@ public boolean awaitMetadataUpdate(long timeout) {
         int version = this.metadata.requestUpdate();
         do {
             poll(timeout);
+            this.metadata.maybeHandleNonRetriableAuthenticationExceptions();
```

Comment:
Hmm.. What happens for metadata updates that occur without an explicit call to `awaitMetadataUpdate`? It could be triggered by the `NetworkClient` itself after all. Maybe we need to add something to the consumer's implementation of `Metadata.Listener`?

Reply:
@hachikuji The scope of the JIRA and this PR are limited to a specific scenario - provide better feedback to applications when security is misconfigured. We currently retry forever, instead we want to throw an `AuthenticationException`. To do this, the first request initiated by any public API needs to process authentication exceptions. And I think that is covered by this PR for producers and consumers.
There is a second scenario, which we do want to handle eventually, but is not addressed by this PR. And that is addressing proper handling of credential updates. If a credential is removed, then we could see authentication exception from any request and we want to propagate this properly. This needs quite extensive testing and is less likely to be hit by users so it is not being addressed now.

Reply:
Btw, we can perhaps do that in a separate PR, but the AdminClient also needs to handle this correctly (not just Producer and Consumer). Also, we should file a JIRA for Streams and Connect to verify that they behave correctly too.

Reply:
@ijuma Yes, I had asked Vahid to do `AdminClient` in a separate PR so that we can get this one committed. He started on it yesterday and has some initial code for review. For Connect and Streams, I will file a separate JIRA. There is another missing piece which is the transactional API with producers, I will do that in another PR now.

Reply:
@vahidhashemian Hmm.. For manual partition assignment, there is no guarantee that the `ensureCoordinatorReady` path will even be executed. We intentionally avoid looking up the coordinator if the user manually sets their own offsets and does not use the commit APIs. Maybe we can add an integration test for this path to ensure that the `AuthenticationException` is somehow getting propagated. Basically it should turn off auto-commit, assign a partition, seek to offset 0 (say), and then call `poll()`.

Reply:
@vahidhashemian It should be easy to add the scenario @hachikuji has suggested above by adding another test similar to `SaslClientsWithInvalidCredentialsTest.testManualAssignmentConsumerWithAuthenticationFailure`.

Reply:
Sure, I'll try to add that scenario. Thanks.